File: Common/TournamentService.m
//
//  TournamentService.m
//  td
//
//  Created by Ryan Drake on 6/28/15.
//  Copyright (c) 2015 HDna Studio. All rights reserved.
//

#import "TournamentService.h"

@interface TournamentService () <TournamentConnectionDelegate>

@property (nonatomic, strong) NSDictionary* dict;

@end

@implementation TournamentService

// initialize with either a NetService, address/port, or unix socket
- (instancetype)initWithNetService:(NSNetService*)service {
    if (self = [super init]) {
        _dict = @{@"service":service, @"name":[service name]};
    }
    return self;
}

- (instancetype)initWithAddress:(NSString*)address andPort:(NSInteger)port {
    if (self = [super init]) {
        _dict = @{@"address":address, @"port":@(port), @"name":[NSString stringWithFormat:@"%@:%li", address, (long)port]};
    }
    return self;
}

- (instancetype)initWithUnixSocket:(NSString*)path {
    if (self = [super init]) {
        _dict = @{@"path":path, @"name":[path lastPathComponent]};
    }
    return self;
}

- (BOOL)isRemote {
    return [self dict][@"path"] == nil;
}

- (NSString*)name {
    return [self dict][@"name"];
}

// retrieve the netService, if applicable
- (NSNetService*)netService {
    return [self dict][@"service"];
}

@end

@implementation TournamentConnection (TournamentService)

- (BOOL)connectToTournamentService:(TournamentService*)service {
    NSNetService* netService = [service dict][@"service"];
    NSString* path = [service dict][@"path"];
    NSString* address = [service dict][@"address"];
    NSNumber* port = [service dict][@"port"];

    if(netService) {
        return [self connectToNetService:netService];
    }

    if(path) {
        return [self connectToUnixSocketNamed:path];
    }

    if(address) {
        return [self connectToAddress:address andPort:[port integerValue]];
    }

    return NO;
}

@end

File: tournamentd/tournament_service.cpp
#include "tournament_service.hpp"

struct tournament_service::impl
{
    std::string path;
    std::string address;
    int port;
    std::string name;
};

// construct from address and port
tournament_service::tournament_service(const std::string& address, int port) : pimpl(new impl)
{
    this->pimpl->address = address;
    this->pimpl->port = port;
    this->pimpl->name = address + ":" + std::to_string(port);
}

// construct from unix socket path
tournament_service::tournament_service(const std::string& path) : pimpl(new impl)
{
    this->pimpl->path = path;
    this->pimpl->name = path.substr(path.find_last_of("/\\") + 1);
}

// default move constructor
tournament_service::tournament_service(tournament_service&& other) = default;

// default destructor
tournament_service::~tournament_service() = default;

// is this a remote service?
bool tournament_service::is_remote() const
{
    return this->pimpl->path.empty();
}

// service name
std::string tournament_service::name() const
{
    return this->pimpl->name;
}
